### PR TITLE
[Examples] create model with custom config on the fly

### PR DESCRIPTION
**Pull request by stas00**
This PR is addressing a need to:
1. be able to quickly whip up a model of any desired size for the big-science experiments.
2. be able to activate gradient checkpointing (later addition)

We already have the functionality to create a model instead of using a pretrained one, but there was no way to control its config - it would choose the defaults of the Config object, which is very doubtful is of any practical use.

This PR:

1. adds a new `PretrainedConfig` method: `update_from_string` so one can update from a string.
```
config.update_from_string("n_embd=10,n_head=5,scale_attn_weights=false,summary_type=super_cls_index")
```

   plus test.

2. adds a new `ModelArguments` arg: `--config_overrides="n_embd=1024,n_head=16,n_layer=48,n_positions=102"` which overrides the default config

3. auto-logs the resulting model size e.g.:
```
Training new model from scratch - Total size=626.69M params
```

Usage:
```
PYTHONPATH=src python examples/pytorch/language-modeling/run_clm.py --dataset_name \
"stas/openwebtext-10k" --output_dir output_dir --overwrite_output_dir --do_train --do_eval \
--max_train_samples 10000 --max_eval_samples 1000 --per_device_train_batch_size 4 \
--per_device_eval_batch_size 4 --num_train_epochs 1 --warmup_steps 8 --block_size 64 --fp16 \
--report_to none --model_type gpt2 --tokenizer_name gpt2 \
--config_overrides "n_embd=1024,n_head=16,n_layer=48,n_positions=1024"
```

Only `run_clm.py` for this experiment.

@sgugger, @LysandreJik

### CODE REVIEW

File: examples/pytorch/language-modeling/run_clm.py [hunk 1]
```diff
@@ -101,6 +105,13 @@ class ModelArguments:
         },
     )
 
+    def __post_init__(self):
+        # convert "key1=val1,key2=val2" into a kwargs dict
+        if self.config_overrides is not None:
+            d = dict(x.split("=") for x in self.config_overrides.split(","))
+            # XXX: need more intelligent code to handle various types of numbers - currently just int
+            self.config_overrides = {k: int(v) for k, v in d.items()}
```
Comment (sgugger) [inline examples/pytorch/language-modeling/run_clm.py:113]: The case should probably be done later on after instantiating a config and checking the type of all values to use the proper type?

File: examples/pytorch/language-modeling/run_clm.py [hunk 2]
```diff
@@ -279,6 +290,9 @@ def main():
     else:
         config = CONFIG_MAPPING[model_args.model_type]()
         logger.warning("You are instantiating a new config instance from scratch.")
+        if model_args.config_overrides is not None:
+            print("Overriding config:", model_args.config_overrides)
+            config.update(model_args.config_overrides)
```
Comment (sgugger) [inline examples/pytorch/language-modeling/run_clm.py:295]: Here it would be easier to cast each value to the right type I think.

**Review by sgugger — APPROVED** [review]
It's indeed not too much, great work! Once ready we should copy this to the other language modeling scripts (where users mostly train from scratch).

File: examples/pytorch/language-modeling/run_clm.py (hunk 2 above)
Comment (stas00) [reply examples/pytorch/language-modeling/run_clm.py]: I added a new method in configuration_utils - so now anybody can use it.

File: src/transformers/configuration_utils.py [hunk 3]
```diff
@@ -667,7 +667,43 @@ def update(self, config_dict: Dict[str, Any]):
...
         """
         for key, value in config_dict.items():
             setattr(self, key, value)
+
+    def update_from_string(self, update_str: str):
+        """
+        Updates attributes of this class with attributes from ``update_str``.
+
+        The expected format is in the form of this example ``key1=10,key2=0.2,key3=true,key4=string`` - i.e. just write
+        ints and floats and strings as is, and for booleans use ``true`` or ``false``.
```
Comment (stas00) [inline src/transformers/configuration_utils.py]: If you like it let's discuss if we want perhaps `;` instead of `,` or anything else before we release it.

Should we expect any `,` in the values of the config object?

And at the moment it only support top-level key updates. so no nested dicts.

File: src/transformers/configuration_utils.py [hunk 4]
```diff
@@ -667,7 +667,43 @@ def update(self, config_dict: Dict[str, Any]):
...
+        """
+
+        d = dict(x.split("=") for x in update_str.split(","))
+        for k, v in d.items():
+            if not hasattr(self, k):
+                raise ValueError(f"key {k} isn't in the original config dict")
+
+            old_v = getattr(self, k)
+            if isinstance(old_v, bool):
+                if v.lower() == "true":
```
Comment (sgugger) [inline src/transformers/configuration_utils.py:696]: Since we're going at lengths to parse this, maybe add all truthy values like y, yes and 1?

File: src/transformers/configuration_utils.py [hunk 5]
```diff
@@ -667,7 +667,43 @@ def update(self, config_dict: Dict[str, Any]):
...
+                    v = True
+                elif v.lower() == "false":
```
Comment (sgugger) [inline src/transformers/configuration_utils.py:698]: Same here.

File: src/transformers/configuration_utils.py [hunk 6]
```diff
@@ -667,7 +667,43 @@ def update(self, config_dict: Dict[str, Any]):
...
+                    v = False
+                else:
+                    raise ValueError(f"can't derive true or false from {v} (key {k})")
+            elif isinstance(old_v, int):
+                v = int(v)
+            elif isinstance(old_v, float):
+                v = float(v)
+            elif not isinstance(old_v, str):
+                raise ValueError(f"{v} (key {k}) is not int/float/bool/str")
```
Comment (sgugger) [inline src/transformers/configuration_utils.py:707]: Let's make this a bit better:
```suggestion
                raise ValueError(f"You can only update int, float, bool or string values in the config, got {v} for key {k}")
```

File: src/transformers/configuration_utils.py (hunk 3 above)
Comment (sgugger) [reply src/transformers/configuration_utils.py]: Some config have lists values sometimes (though it's pretty rare), but then we can imagine parsing on space, so , is fine with me. I think we're good with just the focus on int/float/bool/str.

File: examples/pytorch/language-modeling/run_clm.py [hunk 7]
```diff
@@ -292,7 +285,7 @@ def main():
         logger.warning("You are instantiating a new config instance from scratch.")
         if model_args.config_overrides is not None:
             print("Overriding config:", model_args.config_overrides)
-            config.update(model_args.config_overrides)
+            config.update_from_string(model_args.config_overrides)
```
Comment (sgugger) [inline examples/pytorch/language-modeling/run_clm.py:291]: Very clean!

**Review by sgugger — APPROVED** [review]
I like the new approach a lot!

**Comment by stas00** [comment]
So now can we activate activation checkpointing with: `--config_overrides "gradient_checkpointing=true,use_cache=False"`

**Comment by sgugger** [comment]
This option does not make any sense for pretrained models: in the best case the user will get an error of weights shape mismatch, in the worst case it will just silently yield crappy results.
Thus, the option does not make sense IMO for scripts not used for training models from scratch, have to check manually but I think it's just the scripts for language-modeling which offer that option, so in this case only document the option in their README.

**Comment by stas00** [comment]
> [...] have to check manually but I think it's just the scripts for language-modeling which offer that option, so in this case only document the option in their README.

by "that option" do you mean "gradient_checkpointing"? If so it's available in 30 models out of 59:
```
$ grep -Irl gradient_checkpointing src/transformers/models/*/modeling* | wc -l
30
$ ls -l src/transformers/models/*/modeling* | egrep -v '(flax|tf)'| wc -l
59
```

**Comment by sgugger** [comment]
No I meant the option of training from scratch. I did double check, and it's only in the LM scripts.

**Comment by stas00** [comment]
Right, and i was talking about documenting ` --config_overrides "gradient_checkpointing=true,use_cache=False"`

**Comment by sgugger** [comment]
Yes, it will be a regular training argument in the future.

**Review by LysandreJik — APPROVED** [review]
Great, this looks useful!

An issue with the current implementation/doc is that with the following line:
```
A model can be created on the fly with the help of `--config_overrides`
```
I'm thinking that when I specify this, a model will be created on the fly with the configuration options I specified; but that's not true if I already gave a `--model_name_or_path` argument: in that case, the configuration overrides will simply get ignored.

I would either rephrase this to something along those lines:
```
When training a model from scratch, configuration values may be overridden with the help of `--config_overrides`
```
or add a warning when both `--config_name` and `--config_overrides` or both `--model_name_or_path` and `--config_overrides` are defined.

Or both rephrasing and logging the warning!

Other than that, LGTM! And cool test!

**Comment by stas00** [comment]
Excellent point, @LysandreJik.

I did both.

Though warning not, assert yes.
